packages/micromark-core-commonmark/dev/lib/label-end.js

```diff
@@ -39,22 +39,33 @@ const referenceCollapsedConstruct = {tokenize: tokenizeReferenceCollapsed}
 /** @type {Resolver} */
 function resolveAllLabelEnd(events) {
   let index = -1
-
+  /** @type {Array<Event>} */
+  const newEvents = []
   while (++index < events.length) {
     const token = events[index][1]
+    if (newEvents.length > 0) {
+      push(newEvents, [events[index]])
+    } else {
+      newEvents.push(events[index])
+    }
 
     if (
       token.type === types.labelImage ||
       token.type === types.labelLink ||
       token.type === types.labelEnd
     ) {
       // Remove the marker.
-      events.splice(index + 1, token.type === types.labelImage ? 4 : 2)
+      const offset = token.type === types.labelImage ? 4 : 2
       token.type = types.data
-      index++
+      // Exit effect can skip the check
+      index += offset + 1
+      if (index < events.length) {
```

Comment:
This looks potentially weird? How often is `index < events.length`, `false`?

This check is also already performed by the loop itself? Looking quickly, it looks like `if (index < events.length) { push(newEvents, [events[index]]) }` can be dropped entirely because the loop would do the same anyway?

Reply:
Beacuse the logic is when skip some events, we need to skip the exit effect as well. So the previous code will increment the index after splice the events. 

If we want to create the new effects we need to push it into newEvents to keep the same logic.


I think the code may a little werid as well. Let me try another implemention.

Reply:
> Beacuse the logic is when skipping some events, we need to skip the exit effect as well. So the previous code will increment the index after splice the events.
> 
> If we want to create the new effects we need to push it into newEvents to keep the same logic.
> 
> I think the code may a little werid as well. Let me try another implementaion.

Reply:
@wooorm Thx for the patience reviewing. I have done some code refactoring based on your previous code review. Could you please help review the updated code when you have time?